File: app/src/vdsensor/esp3/events.py
"""EVENT (packet type 0x04) parser (§2.4)."""

from __future__ import annotations

from dataclasses import dataclass

from .framing import Frame
from .packets import PacketType


@dataclass(frozen=True)
class Event:
    event_code: int
    payload: bytes                  # data bytes after the event code
    opt: bytes


def parse_event(frame: Frame) -> Event:
    if frame.packet_type != PacketType.EVENT:
        raise ValueError(f"not an EVENT frame: type={frame.packet_type:#x}")
    if not frame.data:
        raise ValueError("EVENT frame has empty data")
    return Event(event_code=frame.data[0], payload=frame.data[1:], opt=frame.opt)

File: app-rs/src/esp3/events.rs
//! EVENT (packet type 0x04) parser (§2.4).

use super::framing::Frame;
use super::ParseError;

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Event {
    pub event_code: u8,
    pub payload: Vec<u8>,
    pub opt: Vec<u8>,
}

/// Parse an EVENT frame.
///
/// Stub for R1.red.
pub fn parse_event(_frame: &Frame) -> Result<Event, ParseError> {
    unimplemented!("parse_event — implemented by R1.green")
}
